Update ExponentImagePicker.web.ts

isthaison opened the pull request:
add allowsMultipleSelection for web

# Why


# How
return ImagePickerResult[] when allowsMultipleSelection =true

# Test Plan

isthaison committed:
Update ExponentImagePicker.web.ts

add allowsMultipleSelection for web

isthaison committed:
Update ExponentImagePicker.web.ts

barthap commented:
You should also run `yarn build` in module directory to generate `build` files

barthap commented on packages/expo-image-picker/src/ExponentImagePicker.web.ts:
```diff
@@ -92,44 +91,23 @@ function openFileBrowserAsync({
+          const list: ImagePickerResult[] = []
+          for (let i = 0; i < input.files.length; i++) {
+            list.push(await readFile(input.files[i]))
+
           }
-        };
-        // Read in the image file as a binary string.
-        reader.readAsDataURL(targetFile);
+          resolve(list)
```
I'd rather use `Promise.all()` here

barthap replied:
I'd rather use Promise.all() here

bbarthec replied:
Just as @barthap suggested go with sth like this:
```suggestion
resolve(Promise.all(input.files.map(readFile)));
```

isthaison commented:
I got it. I 'll update the patch on request. Best thank's

Vào 17:44, T.2, 27 Th7, 2020 Bartłomiej Klocek <notifications@github.com>
đã viết:

> *@barthap* commented on this pull request.
> ------------------------------
>
> In packages/expo-image-picker/src/ExponentImagePicker.web.ts
> <https://github.com/expo/expo/pull/9402#discussion_r460803402>:
>
> > +          const list: ImagePickerResult[] = []
> +          for (let i = 0; i < input.files.length; i++) {
> +            list.push(await readFile(input.files[i]))
> +
>            }
> -        };
> -        // Read in the image file as a binary string.
> -        reader.readAsDataURL(targetFile);
> +          resolve(list)
>
> I'd rather use Promise.all() here
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/expo/expo/pull/9402#pullrequestreview-455670866>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFNPU6X2GUVA3J7K3F6UW6TR5VK7TANCNFSM4PHZMW5A>
> .
>

bbarthec commented on packages/expo-image-picker/src/ExponentImagePicker.web.ts:
```diff
@@ -92,44 +91,23 @@ function openFileBrowserAsync({
-            };
-          } else {
-            returnRaw();
+        if (allowsMultipleSelection === false) {
```
```suggestion
        if (!allowsMultipleSelection) {
```

bbarthec commented on packages/expo-image-picker/src/ExponentImagePicker.web.ts:
```diff
@@ -141,3 +119,50 @@ function openFileBrowserAsync({
+    reader.onerror = () => {
+      reject(new Error(`Failed to read the selected media because the operation failed.`));
+    };
+    reader.onload = ({ target }) => {
+      const uri = (target as any).result;
+
+      const returnRaw = () => {
+        resolve({
+          cancelled: false,
+          uri,
+          width: 0,
+          height: 0,
+        });
+      };
+      if (typeof target?.result === 'string') {
+        const image = new Image();
+        image.src = target.result;
+        image.onload = function () {
+          resolve({
+            cancelled: false,
+            uri,
+            width: image.naturalWidth ?? image.width,
+            height: image.naturalHeight ?? image.height,
+          });
+        };
+        image.onerror = () => {
+          returnRaw();
+        };
+      } else {
+        returnRaw();
+      }
+    };
+    reader.readAsDataURL(targetFile);
+
+  })
+
+}
+
+
+
```
Some styling suggestions:
```suggestion

async function readFile(targetFile: Blob): Promise<ImagePickerResult> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onerror = () => {
      reject(new Error(`Failed to read the selected media because the operation failed.`));
    };
    reader.onload = ({ target }) => {
      const uri = (target as any).result;
      const returnRaw = () => resolve({
        cancelled: false,
        uri,
        width: 0,
        height: 0,
      });
    
      if (typeof target?.result === 'string') {
        const image = new Image();
        image.src = target.result;
        image.onload = () => resolve({
          cancelled: false,
          uri,
          width: image.naturalWidth ?? image.width,
          height: image.naturalHeight ?? image.height,
        });
        image.onerror = () => returnRaw();
      } else {
        returnRaw();
      }
    };

    reader.readAsDataURL(targetFile);
  })
}
```

bbarthec reviewed (changes requested):
Looks good to me 👍 
Another thing that is missing (apart from rebuilding the `build` folder) is the `CHANGELOG` entry - please add one 😉

isthaison committed:
Update ExponentImagePicker.web.ts

isthaison committed:
Revert "Update ExponentImagePicker.web.ts"

This reverts commit 29dbfb561d6360a33d5486abec1099f8357d602a.

isthaison committed:
Update packages/expo-image-picker/src/ExponentImagePicker.web.ts

pull

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update packages/expo-image-picker/src/ExponentImagePicker.web.ts

pull

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update ExponentImagePicker.web.ts

isthaison committed:
Merge branch 'patch-2' of https://github.com/isthaison/expo into patch-2

isthaison committed:
Update CHANGELOG.md

isthaison committed:
Merge branch 'master' into patch-2

barthap commented on packages/expo-image-picker/CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@
 ### 🐛 Bug fixes
 
 - Fixed downsizing cropped image, when `allowsEditing` was `true`. ([#9316](https://github.com/expo/expo/pull/9316) by [@barthap](https://github.com/barthap))
+- [Web] Return Array `ImagePickerResult` when `allowsMultipleSelection = true` by [@isthaison](https://github.com/isthaison)) 
```
```suggestion
- Return array of `ImagePickerResult` when `allowsMultipleSelection` is set to `true` on Web. ([#9402](https://github.com/expo/expo/pull/9402) by [@isthaison](https://github.com/isthaison)) 
```
I think this is more human-readable.

barthap reviewed (changes requested):
Did you forget to run `yarn build` again? Build files are still missing.

isthaison committed:
Update packages/expo-image-picker/CHANGELOG.md

greate!

Co-authored-by: Bartłomiej Klocek <barthap10@gmail.com>

isthaison commented:
![image](https://user-images.githubusercontent.com/22739578/88571129-58e78c00-d067-11ea-9dd9-91803ac9dbc0.png)
I proceeded to build the file .js but I get error. have you refer for me.

barthap commented:
> env dev: `windows 10`

I think this command may not work on Windows. Try using `tsc` or `npx tsc` directly instead of `yarn build`.

isthaison committed:
update build

isthaison committed:
Delete ExponentImagePicker-test.web.js

isthaison committed:
update

update

isthaison committed:
Update ExponentImagePicker.web.ts

isthaison committed:
update

update

isthaison committed:
Revert "update"

This reverts commit 17cb4468f1b90a4f14ba8183a0ac00136bd6aa61.

lukmccall reviewed (changes requested):
It looks good, hovewer we need to change the return type in the `ImagePicker.ts` too. But when we set it to an `ImagePickerResult | ImagePickerResult[]`, then people always need to check if they get an array or not. I don't like it. Especially that Android and iOS don't support multi-selection. 

So, we could return two different types based on the `allowsMultipleSelection` value. When it is set to:
- `false | undefined` - returns `ImagePickerResult` 
- `true` - returns `ImagePickerMultiResult` which looks like this:
```ts
export type ImagePickerMultipleResult = { cancelled: true } | ({ cancelled: false; selected: ImageInfo[]});
```

To do this, we should write multiple function signatures - [see this demo](https://www.staging-typescript.org/play?jsx=3#code/PTAEEkFsEMHMFMAKBLAxga3gJwHQBcBPAB3gGd9SAoeADyIHss9RCSIYFwA7AM3tAC8oAN6VQoAK5ZkALlCk80rrADcY0AHdkAEzwALOVwmQARtjXi98ZLD15Dxs1gsti8APxyA5Mg7wvoAA+oF4Abjrw9F4utMg8niKgANqYBHIKSrAAunLQXASgAL4uJtCk8ABsACwJGcjKasWU1HSMzKzw7HBIaJhYAPJEeMj0XKSCIurQADbT9BqkAKLayMPKCSb09NPweS5lJKh4CUlGptgANKBnTlkuAI4SM6sECTfmU7PzpACyEtPDIg7ADK8B2RxGXA2Wx2e3UsXick2212XBKZUqNSRMNRjTULQYTFcbCg3RQGGwACUyP9mEJhKBUHlUGCdto5IoJJ1CkFQAAKBlMrgs2bwdmgHgzcpFUAAMi6nF49AAlPiQBAWHp6uhNJ1SHp6P9tKBoNpjSseDwsPAuO03KA4rrQNa8FIuKBILTkEDOr5uqRgOFtJEqOqAGKMUC0aCQH1XVaMw3TY1zeg66bITCa5CkGQEtrEzqkhDkvp-AHenbU0i0iaC5mssUcrBcmXBAWMhui8WS6blFTyMHwI5NhXwbh8JJZQqq5rq7igRatLg2vDFnoU3AaeAmCiUdXbzvul1u8YWq2rwukK6laWjTWdIjQLAx+B4bCgUIzLnNWiE5hlAQwoShIwrDPeADqiwAEIAPrTNAoGoHo64ADLICYL5YAQACCpBAagArqOIMxzAs5aAiCQ4QqMlCFHI66ltggzgWMcoiKR3wUZW8CguCrEJL20oznIiBYPQkA5vAAA8jG9FSNIAgAfPif4FoBwE8EhrGgFBcEIUhKF+OhmHPrh+HCkR4gkV85Fej6fHDqxdEMX4TEDEMkLjPKwicXZFYOdRrHNlyImgGJElSbJbnyVg3E+tWtIqXOYAQZ0K5itm4x4PwMBZjW1qatAzBMrMoAAAZ6fBiHCkZ3QmVh5kEXyyrlZoqwGhIAH5CaWCwMYq7jFoZWoBJRDIDs+ZEhpqAgWBkK6TB1WGWhGGNXhzXCIU7jKqJ4mSeU0VkrFiXKWqYB5MaViFYeprGlp833ps2gEFNAEWbND00e6VUGbVq2mdhG2WaI1kmrZvz2VR-ELUIQnwC5Y7uSxXm7eF+1RXJm6ncwwRY2WUPwDjSmTGD6rXZ0sD0GQoBzLAaCaFqyHyJ1ybOm+boVfjCk1gCbWRuV3NxYTOPlcR7Oulg7qg2D1lCiKbIhQjYOFHRKWgAAKmQzDlN9oCUKIo1jO02uwQAjBMv01chAPrR9LVqEbCgsKbABMltLX9NvGWtZnA4RMs2WRkMBdDTmQkrdGzk7JsKLBADMHv6dbdUIA1fv24H4PB-FYffXI8NR47ozO++cdVEny3-T7gNNSDM6NOrC5C3ub0mh9c1617qfwOnQP2+L9CeSXrnHZuKMl+xvkQ7nvFBZCglStylBoxFB0yULxOqa000d19Ond7bGfNYPw9jKPJaxRPbE+X5IeUXPMOjErqurxjh1C7PW+-jv70EZ3B8U5H37ifMGQ9WK5iRlfM+4x6Svz2pFD+MVsaKVxlAzcX9UEkyzieKWi1k4rRrnbZq4DUaN33GALWztdY6QNpQGOLs44WyEFbQh9VfYgMstHEusc8CwXdiwz2QCiHHxBuLO+s9HL5xYC2BGDd6E8MYXwxOgiCHV3YbXf2VlrISMJlI4KEol5FwUcbJRsEK6qKrt7DRxD66zlVs0RxQA). More information you can find [here](https://stackoverflow.com/a/52818072/13361748).

Unfortunately, it's a breaking change... I think. Hovewer, for me, it's worth it to do ;)
What do you think? 
cc @bbarthec, @barthap

barthap commented:
:+1:  for the solution from demo. It looks really clean. However, [I modified it a bit](https://www.staging-typescript.org/play?#code/PTAEEkFsEMHMFMAKBLAxga3gJwHQBcBPAB3gGd9SAoeADyIHss9RCSIYFwA7AM3tAC8oAN6VQoAK5ZkALlCk80rrADcY0AHdkAEzwALOVwmQARtjXi98ZLD15Dxs1gsti8APxyA5Mg7wvoAA+oF4Abjrw9F4utMg8niKgANqYBHIKSrAAunLQXASgAL4uJtCk8ABsACwJGcjKasWU1HSMzKzw7HBIaJhYAPJEeMj0XKSCIurQADbT9BqkAKLayMPKCSb09NPweS5lJKh4CUlGptgANKBnTlkuAI4SM6sECTfmU7PzpACyEtPDIg7ADK8B2RxGXA2Wx2e3UsXick2212XBKZUqNSRMNRjTULQYTFcbCg3RQGGwACUyP9mEJhKBUHlUGCdto5IoJJ1CkFQAAKBlMrgs2bwdmgHgzcpFUAAMlAizoeW0AB5SZxePQAHwASnxIAgLD09XQmk6pD09H+2lA0G0NpWPB4WHgXHablAcTNoBdeCkXFAkFpyCBnV83VIwHC2kiVANADFGKBaNBIKGrqtGVbpja5vRTdNkJgjchSDICW1iZ11T0KVg-gCQztqaRaRNBczWWKOVguTLggLGZ3ReLJdNyip5GD4Edu10NXwkllCnrmgbuArWlxXXga+S+jgNPATBRKAaj0OA77-eNHc6d1XSFdStLRkbOkRoFhU-A8NhQKEMxcs0tCEswZQEMKEoSMKwxvgA6osABCAD60zQDBqB6DWAAyyAmN+WAEAAgqQkGoAK6jiDMcwLA2gIgtOEKjJQhRyHuvTYIMcFjHKIg0d89FNvAoLgjxCRjtKK5yIgWD0JApbwCqiqflwqocXWLa0lqWr4qBlYQVBPCYTxoCIah6GYdhfh4QRX4kWRwqUeI1FfHRwahqJM48ax7F+PuXFDJC4zysIAnuY2nlMTxPZctJoCyfJinKUqalqv5nH1h5zY0gCOn6mA8GdNuYoluMeD8DAxati6RrQMwTKzKAAAG5loRhwrWd0tmEQ55F8jqzWaKsloSOB+S2lgsDGDu4xaE1qDyUQyA7BWRKGag0GwZCZnIe1Vm4fhvWkf1wiFO4OoyXJCnlClqnqRlmm5Xg+VrmAyqgFYtUXnaNrGdtb6bNoBBreBjmbf9zEBm1lmdYddlESdTmiC5tpub82UidFO1CJJ8C+fOtZ9NxwWXQl13JSpyrpWSmVaXlvJU2lGl9EJob0y9WqTKjBpfZ0sD0GQoBzLAaCaMaWHyKNOY+r+-otSzVLPUNSbNYrWWRTlrYAs1VGy36WABijqMuUKIpsrF+Oo4UrFvaAAAqZDMOUUOgJQoiLWM7ROyhACMEzQBo0CZjDHVYfDx3gwNaiewoLA+wATAHQch3tsPhzZR32UjFHG65tEY5rWNiZClusausfewoKEAMzJ8HzChwdmcI31yN6+FhcMcX3mlxKUr4yuMejHHf7V1U9epxZYddQgPXZ1HZ2rjbZ5gBu6unqDtrg1trvp7P8Dz4jUd6-QQUj35tN1iTI98WF6Ns4xJejBJA9FJQZOJTdSlMw9V99BzV6W8Nq71MvvCOC9+qn3PmMS+CAAoDBgSFfiD9MZeShpbG2n8Ka3V-jTeBmVH7wEAbpECrR1o70hmAmeEDj5QNRmfHiZZCYIJvrxekWCrpJVwalP+BCnraxeozXh+CibYCISQyYetryG12tPZu3Us50Kcow0mjQ7aOzji7Uy7tKCV3jtXf2QhA4NzkftOGLdI79QriPKueAUJJ2MSnRuacaGWMge3VGnciHoJiiwXsg8bFewMfYuuTjTFNwsYo1uOdnIuW8Wg7Gow5B43LsPYJY97ET3CVPcxGdolWORkPW2q8HZWGlHoaAGBxjhRYPwUCH1twKFKh0cY9QqyfXqHgOMYByidDsHgIgZYQAKCqegegoRsA8FojgRakBgCPCdsFYAFQAAMAB2AAnAAVgTtsiowBWmoGkEMAAtB0U5kBsCi2UJQDo5NfCrGQJMiYdRlC8neFgXkyJYQBmCCYGwXTeRkVMNsD5-xpi8hgjGHg9QxRqHuXg+2XMhD22TDQP8alxiyUecMF57gHagDkAyJIABpT0AZUj0B4A7HIDsyVZCKCoIAA) and added `Expand<T>` hack, which expands objects in type hints. Compare both demos and check, which _type hints_ are more readable.

> Unfortunately, it's a breaking change... I think. Hovewer, for me, it's worth it to do ;)

IMO it won't be a breaking change - when `allowsMultipleSelection` is `false`, return type won't change compared to previous versions, and version with `true` didn't even work at all.

isthaison commented:
I got it. We'll update it soon 🌼🌼

Vào 17:59, T.3, 28 Th7, 2020 Bartłomiej Klocek <notifications@github.com>
đã viết:

> 👍 for the solution from demo. It looks really clean. However, I modified
> it a bit
> <https://www.staging-typescript.org/play?#code/PTAEEkFsEMHMFMAKBLAxga3gJwHQBcBPAB3gGd9SAoeADyIHss9RCSIYFwA7AM3tAC8oAN6VQoAK5ZkALlCk80rrADcY0AHdkAEzwALOVwmQARtjXi98ZLD15Dxs1gsti8APxyA5Mg7wvoAA+oF4Abjrw9F4utMg8niKgANqYBHIKSrAAunLQXASgAL4uJtCk8ABsACwJGcjKasWU1HSMzKzw7HBIaJhYAPJEeMj0XKSCIurQADbT9BqkAKLayMPKCSb09NPweS5lJKh4CUlGptgANKBnTlkuAI4SM6sECTfmU7PzpACyEtPDIg7ADK8B2RxGXA2Wx2e3UsXick2212XBKZUqNSRMNRjTULQYTFcbCg3RQGGwACUyP9mEJhKBUHlUGCdto5IoJJ1CkFQAAKBlMrgs2bwdmgHgzcpFUAAMlAizoeW0AB5SZxePQAHwASnxIAgLD09XQmk6pD09H+2lA0G0NpWPB4WHgXHablAcTNoBdeCkXFAkFpyCBnV83VIwHC2kiVANADFGKBaNBIKGrqtGVbpja5vRTdNkJgjchSDICW1iZ11T0KVg-gCQztqaRaRNBczWWKOVguTLggLGZ3ReLJdNyip5GD4Edu10NXwkllCnrmgbuArWlxXXga+S+jgNPATBRKAaj0OA77-eNHc6d1XSFdStLRkbOkRoFhU-A8NhQKEMxcs0tCEswZQEMKEoSMKwxvgA6osABCAD60zQDBqB6DWAAyyAmN+WAEAAgqQkGoAK6jiDMcwLA2gIgtOEKjJQhRyHuvTYIMcFjHKIg0d89FNvAoLgjxCRjtKK5yIgWD0JApbwCqiqflwqocXWLa0lqWr4qBlYQVBPCYTxoCIah6GYdhfh4QRX4kWRwqUeI1FfHRwahqJM48ax7F+PuXFDJC4zysIAnuY2nlMTxPZctJoCyfJinKUqalqv5nH1h5zY0gCOn6mA8GdNuYoluMeD8DAxati6RrQMwTKzKAAAG5loRhwrWd0tmEQ55F8jqzWaKsloSOB+S2lgsDGDu4xaE1qDyUQyA7BWRKGag0GwZCZnIe1Vm4fhvWkf1wiFO4OoyXJCnlClqnqRlmm5Xg+VrmAyqgFYtUXnaNrGdtb6bNoBBreBjmbf9zEBm1lmdYddlESdTmiC5tpub82UidFO1CJJ8C+fOtZ9NxwWXQl13JSpyrpWSmVaXlvJU2lGl9EJob0y9WqTKjBpfZ0sD0GQoBzLAaCaMaWHyKNOY+r+-otSzVLPUNSbNYrWWRTlrYAs1VGy36WABijqMuUKIpsrF+Oo4UrFvaAAAqZDMOUUOgJQoiLWM7ROyhACMEzQBo0CZjDHVYfDx3gwNaiewoLA+wATAHQch3tsPhzZR32UjFHG65tEY5rWNiZClusausfewoKEAMzJ8HzChwdmcI31yN6+FhcMcX3mlxKUr4yuMejHHf7V1U9epxZYddQgPXZ1HZ2rjbZ5gBu6unqDtrg1trvp7P8Dz4jUd6-QQUj35tN1iTI98WF6Ns4xJejBJA9FJQZOJTdSlMw9V99BzV6W8Nq71MvvCOC9+qn3PmMS+CAAoDBgSFfiD9MZeShpbG2n8Ka3V-jTeBmVH7wEAbpECrR1o70hmAmeEDj5QNRmfHiZZCYIJvrxekWCrpJVwalP+BCnraxeozXh+CibYCISQyYetryG12tPZu3Us50Kcow0mjQ7aOzji7Uy7tKCV3jtXf2QhA4NzkftOGLdI79QriPKueAUJJ2MSnRuacaGWMge3VGnciHoJiiwXsg8bFewMfYuuTjTFNwsYo1uOdnIuW8Wg7Gow5B43LsPYJY97ET3CVPcxGdolWORkPW2q8HZWGlHoaAGBxjhRYPwUCH1twKFKh0cY9QqyfXqHgOMYByidDsHgIgZYQAKCqegegoRsA8FojgRakBgCPCdsFYAFQAAMAB2AAnAAVgTtsiowBWmoGkEMAAtB0U5kBsCi2UJQDo5NfCrGQJMiYdRlC8neFgXkyJYQBmCCYGwXTeRkVMNsD5-xpi8hgjGHg9QxRqHuXg+2XMhD22TDQP8alxiyUecMF57gHagDkAyJIABpT0AZUj0B4A7HIDsyVZCKCoIAA>
> and added Expand<T> hack, which expands objects in type hints. Compare
> both demos and check, which *type hints* are more readable.
>
> Unfortunately, it's a breaking change... I think. Hovewer, for me, it's
> worth it to do ;)
>
> IMO it won't be a breaking change - when allowsMultipleSelection is false,
> return type won't change compared to previous versions, and version with
> true didn't even work at all.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/expo/expo/pull/9402#issuecomment-664974233>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFNPU6X4LY5A4HI4K2JH2BTR52VQTANCNFSM4PHZMW5A>
> .
>

isthaison committed:
Update ImagePicker.types.ts

isthaison committed:
update

update

isthaison committed:
Merge branch 'master' into patch-2

isthaison committed:
upda

upda

isthaison committed:
Merge branch 'patch-2' of https://github.com/isthaison/expo into patch-2

isthaison committed:
update

isthaison committed:
upffa

isthaison committed:
Merge branch 'master' into patch-2

isthaison committed:
Update ImagePicker.types.ts

isthaison committed:
Merge branch 'patch-2' of https://github.com/isthaison/expo into patch-2

isthaison committed:
Merge branch 'patch-2' of https://github.com/isthaison/expo into patch-2

isthaison committed:
update

update

isthaison committed:
update

update

isthaison commented:
Hi @barthap .
cc @bbarthec  , @lukmccall  
We still have a way to return the corresponding value

````
// ImagePicker.types.ts
export type ExpandImagePickerOptions =
  | (ImagePickerOptions & { allowsMultipleSelection: true })
  | ImagePickerOptions;

export type ExpandImagePickerResult<T> = T extends ImagePickerOptions & { allowsMultipleSelection: true }
  ? ImagePickerMultipleResult
  : ImagePickerResult

// ImagePicker.ts
export async function launchImageLibraryAsync<T extends ExpandImagePickerOptions>(
  options: T
): Promise<ExpandImagePickerResult<T>> {
  if (!ExponentImagePicker.launchImageLibraryAsync) {
    throw new UnavailabilityError('ImagePicker', 'launchImageLibraryAsync');
  }
  return await ExponentImagePicker.launchImageLibraryAsync(options);
}
````

I think this way
Not nice but it's quite convenient.

bbarthec commented on packages/expo-image-picker/src/ImagePicker.types.ts:
```diff
@@ -38,19 +38,31 @@ export type ImageInfo = {
   exif?: boolean;
   base64?: boolean;
   videoExportPreset?: VideoExportPreset;
+   allowsMultipleSelection?: boolean;
```
```suggestion
  allowsMultipleSelection?: boolean;
```

bbarthec commented on packages/expo-image-picker/src/ImagePicker.types.ts:
```diff
@@ -38,19 +38,31 @@ export type ImageInfo = {
 
+export type ImagePickerMultipleResult = { cancelled: true } | ({ cancelled: false; selected: ImageInfo[] });
+
+
```
```suggestion
```

bbarthec commented on packages/expo-image-picker/src/ImagePicker.types.ts:
```diff
@@ -38,19 +38,31 @@ export type ImageInfo = {
   allowsMultipleSelection: boolean;
 };
+
+
```
```suggestion
```

bbarthec commented on packages/expo-image-picker/src/ImagePicker.types.ts:
```diff
@@ -38,19 +38,31 @@ export type ImageInfo = {
 };
+
+
+export type ExpandImagePickerOptions =
+  | (ImagePickerOptions & { allowsMultipleSelection: true })
+  | ImagePickerOptions;
```
Having the name of this union type sound as it was a function is confusing.
Moreover I don't see any reason to have this union type at all as `ImagePikcerOptions` already defines a property `{ allowsMultipleSelection: boolean }`.
```suggestion
```

bbarthec commented on packages/expo-image-picker/src/ImagePicker.types.ts:
```diff
@@ -38,19 +38,31 @@ export type ImageInfo = {
+  | (ImagePickerOptions & { allowsMultipleSelection: true })
+  | ImagePickerOptions;
+
+export type ExpandImagePickerResult<T> = T extends ImagePickerOptions & { allowsMultipleSelection: true }
+  ? ImagePickerMultipleResult
+  : ImagePickerResult
```
Getting rid of `ExpandImagePickerOptions` would make us write sth like this one:
```suggestion
export type ExpandImagePickerResult<T extends ImagePickerOptions> = T extends { allowsMultipleSelection: true }
  ? ImagePickerMultipleResult
  : ImagePickerResult

```
Here's more contextual example: [typescript playground](https://www.staging-typescript.org/play?#code/KYDwDg9gTgLgBDAnmYcCSBbAhgc2GgOwDMI4BeOAbwF8BuAKHtElgWVU12AAUBLAYwDWwKAHkwMXhAIBnclThYANkogB3GQFkArkslglwAMrBD-SdID8ALjgAjCBENYCtOHUbNo8JCnTY8PiERACVgGV14Cko4fhd+U0MAE1sYKG1UajgAHzgAChi4ggSVYBS4ImUZTLgAMn8uQhIASgYvVl8OAJ4BYSgdPV4DYDCIvXlC+MSy1PSa3ILYqdLyyqVqt2qzGBmGvCaIAG0AXWpWz3BvNj8AUXAXJM5A3tDwyIAeABU4UB2CJLkTx6wTEEiksgAfPJvr9gP85DFlKoNAN9IYTNtwbMMu56HA4JY9sC+qihoZRpE8XBbECgn0KXpGExLqwsDJEMUKtpihYCHAlFhufwABZAgAyvDsUCwUEQAEF2cUvj8QH8AUS6SJxLyZBC8hAwdIZLZPs1bNwoBAMLxqu87mAHrSXlAGTAvhCoZQqQB6b1wDDAGDCiBJewhxD0aiMX1wT7heDVczguAXFjwIhC3mKGTVWCfdjuvIAfRNzSoUa9+LZuZg+ZQ7ydINdeqwaiwvHgAqFou6EqlMvliv4BTO5yrOZEtYLjfpbz0LbbHf5guKPa4felsoVHOHiJU6i0kTJxlMwCT0ls3KSwCIvAIZXczTH2ZrdeADe6mpdc5gC-bnZXEVxUlTdBx3RYkQPUlhgxM9eVsNZqkfZ9q0nN8Py4L9oPJH8-yXLtV2A-styHCD9xRI8YNPc8CGxTInwYaggA)

bbarthec reviewed (changes requested):
In general I like the idea where this is going.
Please see my suggestion regarding the clean and efficient way of describing the user-facing API.
No matter whether it's a breaking change or not I'd like to go this way. 👍

isthaison committed:
Update packages/expo-image-picker/src/ImagePicker.types.ts

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update packages/expo-image-picker/src/ImagePicker.types.ts

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update packages/expo-image-picker/src/ImagePicker.types.ts

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update packages/expo-image-picker/src/ImagePicker.types.ts

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

bbarthec commented on packages/expo-image-picker/src/ImagePicker.types.ts:
```diff
@@ -38,19 +38,26 @@ export type ImageInfo = {
 };
+
+
+export type ExpandImagePickerResult<T> = T extends ImagePickerOptions & { allowsMultipleSelection: true }
+  ? ImagePickerMultipleResult
+  : ImagePickerResult
```
Your code would allow passing into `ExpandImagePickerResult` any js object or primitive.
What I suggest is to narrow down what's accepted in this mapped type to `ImagePickerOptions` (either having `allowsMultipleSelection: true` or not):
```suggestion
export type ExpandImagePickerResult<T extends ImagePickerOptions> = T extends { allowsMultipleSelection: true }
  ? ImagePickerMultipleResult
  : ImagePickerResult
```

isthaison replied:
i got it );

bbarthec commented on packages/expo-image-picker/src/ImagePicker.ts:
```diff
@@ -45,6 +38,15 @@ export async function launchCameraAsync(
   return await ExponentImagePicker.launchCameraAsync(options);
 }
 
+export async function launchImageLibraryAsync<T extends ExpandImagePickerOptions>(
```
```suggestion
export async function launchImageLibraryAsync<T extends ImagePickerOptions>(
```

bbarthec commented on packages/expo-image-picker/src/ImagePicker.ts:
```diff
@@ -9,6 +9,8 @@ import {
   MediaTypeOptions,
   ImagePickerOptions,
   VideoExportPreset,
+  ExpandImagePickerResult,
+  ExpandImagePickerOptions,
```
```suggestion
  ExpandImagePickerResult,
```

bbarthec commented on packages/expo-image-picker/src/ExponentImagePicker.web.ts:
```diff
@@ -75,7 +78,7 @@ function openFileBrowserAsync({
   capture = false,
   allowsMultipleSelection = false,
-}: OpenFileBrowserOptions): Promise<ImagePickerResult> {
+}: OpenFileBrowserOptions): Promise<ImagePickerResult | ImagePickerMultipleResult> {
```
I'd like to make this methods signature more explicit what's happening, just like you did in `ImagePicker.ts`.
```suggestion
function openFileBrowserAsync<T extends OpenFileBrowserOptions>({
  mediaTypes,
  capture = false,
  allowsMultipleSelection = false,
}: T): Promise<ExpandImagePickerResult<T>> {
```

isthaison committed:
Update packages/expo-image-picker/src/ImagePicker.types.ts

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update packages/expo-image-picker/src/ExponentImagePicker.web.ts

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update packages/expo-image-picker/src/ImagePicker.ts

Co-authored-by: Bartłomiej Bukowski <bartlomiejbukowski.b@gmail.com>

isthaison committed:
Update ImagePicker.ts

isthaison committed:
Update ImagePicker.ts

isthaison committed:
Merge branch 'master' into patch-2

isthaison committed:
update

update

isthaison reviewed (commented):
🔢

bbarthec reviewed (changes requested):
LGTM 👍 
I need you to run `yarn build` in `packages/expo-image-picker` to rebuild the package and make CI green for this PR.

barthap commented:
Closing in favour of #9673 - I rebuilt the JS there
Thank you @isthaison for Your contribution! 👍
